examples/cluster-metrics/member/main.go

```diff
@@ -19,25 +19,25 @@ const (
 )
 
 // Logger is message middleware which logs messages before continuing to the next middleware
-func Logger(next actor.ActorFunc) actor.ActorFunc {
-	fn := func(context actor.Context) {
-		switch context.Message().(type) {
+func Logger(next actor.ReceiverFunc) actor.ReceiverFunc {
+	fn := func(context actor.ReceiverContext, env *actor.MessageEnvelope) {
+		switch env.Message.(type) {
 		case *actor.Started:
-			log.Printf("actor started " + context.Self().String())
+			log.Printf("actor started " + context.(actor.Context).Self().String())
```

Comment:
I got this exact same problem (casting) with the `ReceiverContext` in some of my middleware. Maybe it would make sense to extend the `ReceiverContext` with the some of the properties like `Self`?
I could also be the `SenderMiddleware` should use `Context` instead, since that would allow for more powerful Middleware solutions?

Reply:
@emilingerslev Yep! I also mentioned this comment https://github.com/AsynkronIT/protoactor-go/pull/251#issuecomment-449430686 , this is a very annoying situation. I'm also trying to come up some better ways to handle this situation, but it seems Go doesn't allow duplicate interface property, so it's a bit tricky where to put it = P

Reply:
@PotterDai so I tried to look into this a bit, intrigued by the problem 😊 I see that you mirrored the interfaces for the middleware in the dotnet implementation, so it seems like it intended by @rogeralsing. However I still would like to iterate a bit on this, since it seems like there is something here that could be improved.

About the limitation of go a way to go about it is splitting the context parts into separate interfaces and composition them only in one layer like this 

```go
package actor

import "time"

type Context interface {
	InfoContextPart
	ContextPart
	ReceiverContextPart
	SenderContextPart
}

type SenderContext interface {
	InfoContextPart
	SenderContextPart
}

type ReceiverContext interface {
	InfoContextPart
	ReceiverContextPart
}

type InfoContextPart interface {
	// Parent returns the PID for the current actors parent
	Parent() *PID

	// Self returns the PID for the current actor
	Self() *PID

	// Sender returns the PID of actor that sent currently processed message
	Sender() *PID

	// Actor returns the actor associated with this context
	Actor() Actor
}

// Context contains contextual information for actors
type ContextPart interface {
	// ReceiveTimeout returns the current timeout
	ReceiveTimeout() time.Duration

	// Returns a slice of the actors children
	Children() []*PID

	// Respond sends a response to the to the current `Sender`
	// If the Sender is nil, the actor will panic
	Respond(response interface{})

	// Stash stashes the current message on a stack for reprocessing when the actor restarts
	Stash()

	// Watch registers the actor as a monitor for the specified PID
	Watch(pid *PID)

	// Unwatch unregisters the actor as a monitor for the specified PID
	Unwatch(pid *PID)

	// SetReceiveTimeout sets the inactivity timeout, after which a ReceiveTimeout message will be sent to the actor.
	// A duration of less than 1ms will disable the inactivity timer.
	//
	// If a message is received before the duration d, the timer will be reset. If the message conforms to
	// the NotInfluenceReceiveTimeout interface, the timer will not be reset
	SetReceiveTimeout(d time.Duration)

	CancelReceiveTimeout()

	// Forward forwards current message to the given PID
	Forward(pid *PID)

	AwaitFuture(f *Future, continuation func(res interface{}, err error))
}

type SenderContextPart interface {
	// Message returns the current message to be processed
	Message() interface{}

	// MessageHeader returns the meta information for the currently processed message
	MessageHeader() ReadonlyMessageHeader

	// Send sends a message to the given PID
	Send(pid *PID, message interface{})

	// Request sends a message to the given PID and also provides a Sender PID
	Request(pid *PID, message interface{})

	// Request sends a message to the given PID and also provides a Sender PID
	RequestWithCustomSender(pid *PID, message interface{}, sender *PID)

	// RequestFuture sends a message to a given PID and returns a Future
	RequestFuture(pid *PID, message interface{}, timeout time.Duration) *Future
}

type ReceiverContextPart interface {
	Receive(envelope *MessageEnvelope)
}

type SpawnerContext interface {
	// Spawn starts a new child actor based on props and named with a unique id
	Spawn(props *Props) *PID

	// SpawnPrefix starts a new child actor based on props and named using a prefix followed by a unique id
	SpawnPrefix(props *Props, prefix string) *PID

	// SpawnNamed starts a new child actor based on props and named using the specified name
	//
	// ErrNameExists will be returned if id already exists
	SpawnNamed(props *Props, id string) (*PID, error)
}
```

That would, as far as I can see, work but it's still hard to find the right way to "cut" these contexts.

Reply:
This makes sense! I'll make a change on this one, and on dotnet version as well! Thank you for the proposal! = D

Reply:
@emilingerslev Just implemented in this commit:
https://github.com/AsynkronIT/protoactor-go/pull/251/commits/3ec042d2b3fc5849fe8243397b268780ae58749b

Reply:
Looks good @PotterDai 👍 this should solve quite a few "casts" in the middleware level 😃 And glad you liked the proposal. Was it possible to mimic the same behavior/setup in dotnet?

Reply:
Just created a PR at dotnet version as well, it's much simpler, since it doesn't have the problem of duplicate properties = )